util/stmtsummary/statement_summary.go

```diff
@@ -0,0 +1,213 @@
+	schemaName string
+	digest     string
+	// TODO: add plan digest
+	// `hash` is the hash value of this object
+	hash []byte
+}
+
+// Hash implements SimpleLRUCache.Key
+func (key *stmtSummaryCacheKey) Hash() []byte {
+	if len(key.hash) == 0 {
+		key.hash = make([]byte, 0, len(key.schemaName)+len(key.digest))
+		key.hash = append(key.hash, hack.Slice(key.digest)...)
+		key.hash = append(key.hash, hack.Slice(strings.ToLower(key.schemaName))...)
+	}
+	return key.hash
+}
+
+// A LRU cache that stores statement summary.
+type stmtSummaryByDigest struct {
+	// It's rare to read concurrently, so RWMutex is not needed.
+	sync.RWMutex
+	summaryMap *kvcache.SimpleLRUCache
+}
+
+// StmtSummary is the global object for statement summary.
+var StmtSummary = NewStmtSummaryByDigest()
+
+// Summary of each type of statements.
+type stmtSummary struct {
+	// It's rare to read concurrently, so RWMutex is not needed.
+	sync.Mutex
+	schemaName      string
+	digest          string
+	normalizedSQL   string
+	sampleSQL       string
+	execCount       uint64
+	sumLatency      uint64
+	maxLatency      uint64
+	minLatency      uint64
+	sumAffectedRows uint64
+	// Number of rows sent to client.
+	sumSentRows uint64
+	// The first time this type of SQL executes.
+	firstSeen time.Time
+	// The last time this type of SQL executes.
+	lastSeen time.Time
+}
+
+// StmtExecInfo records execution status of each statement.
+type StmtExecInfo struct {
+	SchemaName    string
+	OriginalSQL   string
+	NormalizedSQL string
+	Digest        string
+	TotalLatency  uint64
+	AffectedRows  uint64
+	// Number of rows sent to client.
+	SentRows  uint64
+	StartTime time.Time
+}
+
+// NewStmtSummaryByDigest creates a stmtSummaryByDigest.
+func NewStmtSummaryByDigest() *stmtSummaryByDigest {
+	maxStmtCount := config.GetGlobalConfig().StmtSummary.MaxStmtCount
+	return &stmtSummaryByDigest{
+		summaryMap: kvcache.NewSimpleLRUCache(maxStmtCount, 0, 0),
+	}
+}
+
+// Convert StmtExecInfo to stmtSummary
+func convertStmtExecInfoToSummary(sei *StmtExecInfo) *stmtSummary {
+	// Trim SQL to size MaxSQLLength
+	maxSQLLength := config.GetGlobalConfig().StmtSummary.MaxSQLLength
+	normalizedSQL := sei.NormalizedSQL
+	if len(normalizedSQL) > int(maxSQLLength) {
+		normalizedSQL = normalizedSQL[:maxSQLLength]
+	}
+	sampleSQL := sei.OriginalSQL
+	if len(sampleSQL) > int(maxSQLLength) {
+		sampleSQL = sampleSQL[:maxSQLLength]
+	}
+
+	return &stmtSummary{
+		schemaName:      sei.SchemaName,
+		digest:          sei.Digest,
+		normalizedSQL:   normalizedSQL,
+		sampleSQL:       sampleSQL,
+		execCount:       1,
+		sumLatency:      sei.TotalLatency,
+		maxLatency:      sei.TotalLatency,
+		minLatency:      sei.TotalLatency,
+		sumAffectedRows: sei.AffectedRows,
+		sumSentRows:     sei.SentRows,
+		firstSeen:       sei.StartTime,
+		lastSeen:        sei.StartTime,
+	}
+}
+
+// Add a new StmtExecInfo to stmtSummary
+func addStmtExecInfoToSummary(sei *StmtExecInfo, ss *stmtSummary) {
```

Comment:
How about that we make this function a method of `stmtSummary` ?

Reply:
1. change the function name to `NewStmtSummary(sei *StmtExecInfo)`
2. `ss := new(stmtSummary); ss.convertFromExecInfo(sei)`
which one?

Reply:
func (ss *stmtSummary) add(sei *StmtExecInfo)

Reply:
Addressed.